### PR TITLE
Refactor error handling in the HMPPS API client

### PR DESCRIPTION
**Pull request by ollietreend**
I've simplified the way errors are handled within our API client (which we use to speak to HMPPS APIs, such as the Prison API, Keyworker API, etc.)

Previously, the client would rescue errors from Faraday (the HTTP client that we use), and raise every error as a generic `HmppsApi::APIError` error. This meant that the true cause of an error was obscured from us, such that we weren't able to determine which HTTP status the error was, or whether it was a 'client error' (4xx) or 'server error' (5xx).

With this refactor, I've removed the `APIError` custom error class, and instead the API client will simply allow Faraday exceptions to bubble up and be caught higher up in the code.

This has also led me to remove the `log404 = true` parameter from the `HmppsApi::Client` initialise method. Since 404 errors now 'bubble up' as `Faraday::ResourceNotFound` errors, it is the responsibility of the caller to either: rescue & handle the error, or allow it to continue bubbling up through the application. All uncaught exceptions will be recorded in Sentry.

This has allowed me to simplify the RSpec tests, because the spec tag `:raven_intercept_exception` is no longer needed. I've also been able to remove the custom exception handler `AllocationManager::ExceptionHandler` as this is no longer used.

In summary:

- Faraday errors will now be raised when an API request fails. This replaces the `HmppsApi::APIError` class.
- Remove unneeded tag `:raven_intercept_exception` from RSpec tests
- Remove unused class `AllocationManager::ExceptionHandler`
- Remove unneeded option `log404` from API client `HmppsApi::Client.new`

### CODE REVIEW

File: app/services/hmpps_api/client.rb [hunk 1]
```diff
@@ -7,9 +7,8 @@ module HmppsApi
   class Client
     APIError = Class.new(StandardError)
```
Comment (starswan) [inline app/services/hmpps_api/client.rb:8]: needs to be removed now?

**Review by starswan — APPROVED** [review]
One little thing needs removing

File: app/services/hmpps_api/client.rb (hunk 1 above)
Comment (ollietreend) [reply app/services/hmpps_api/client.rb]: Thanks @starswan. Removed.
